File: app11_swift_ios/SwiftGuardUI/Sources/SwiftGuardUI/ViewModels/CardSuitViewModel.swift
import Foundation
import SwiftData
import SwiftGuardData

/// Generic, parameterized by suit OR edition — the same "one view model,
/// many routes" pattern app09's suit-archive.php used, backing every
/// suit-browsing user story (US-05–US-12) with one type.
@Observable
@MainActor
public final class CardSuitViewModel {
    public private(set) var cards: [CornucopiaCard] = []
    public private(set) var errorMessage: String?

    private let repository: CardRepository

    public init(modelContext: ModelContext) {
        self.repository = SwiftDataCardRepository(modelContext: modelContext)
    }

    public func loadSuit(_ suitCode: String) {
        do {
            cards = try repository.bySuit(suitCode)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func loadEdition(_ edition: String) {
        do {
            cards = try repository.byEdition(edition)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: app12_kotlin_android/ui/src/main/kotlin/com/kotlinguard/ui/viewmodel/CardSuitViewModel.kt
package com.kotlinguard.ui.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kotlinguard.data.model.CornucopiaCardEntity
import com.kotlinguard.data.repository.CardRepository
import kotlinx.coroutines.launch

class CardSuitViewModel(private val repository: CardRepository) : ViewModel() {
    var cards by mutableStateOf<List<CornucopiaCardEntity>>(emptyList())
        private set

    fun loadSuit(suitCode: String) {
        viewModelScope.launch { cards = repository.bySuit(suitCode) }
    }

    fun loadEdition(edition: String) {
        viewModelScope.launch { cards = repository.byEdition(edition) }
    }
}
